File: lib/tabula/entities/text_chunk.rb
module Tabula
  ##
  # a "collection" of TextElements
  class TextChunk < ZoneEntity
    attr_accessor :font, :font_size, :text_elements, :width_of_space

    ##
    # initialize a new TextChunk from a TextElement
    def self.create_from_text_element(text_element)
      raise TypeError, "argument is not a TextElement" unless text_element.instance_of?(TextElement)
      tc = self.new(*text_element.tlwh)
      tc.text_elements = [text_element]
      return tc
    end

    def self.group_by_lines(text_chunks)
      bbwidth = text_chunks.max_by(&:right).right - text_chunks.min_by(&:left).left

      l = Line.new
      l << text_chunks.first

      lines = text_chunks[1..-1].inject([l]) do |lines, te|
        if lines.last.horizontal_overlap_ratio(te) < 0.01
          # skip lines such that:
          # - are wider than the 90% of the width of the text_chunks bounding box
          # - it contains a single repeated character
          if lines.last.width / bbwidth > 0.9 \
            && l.text_elements.all? { |te| te.text =~  SAME_CHAR_RE }
            lines.pop
          end
          lines << Line.new
        end
        lines.last << te
        lines
      end

      if lines.last.width / bbwidth > 0.9 \
         && l.text_elements.all? { |te| te.text =~ SAME_CHAR_RE }
        lines.pop
      end

      lines.map!(&:remove_sequential_spaces!)
    end

    ##
    # returns a list of column boundaries (x axis)
    # +lines+ must be an array of lines sorted by their +top+ attribute
    def self.column_positions(lines)
      return [] if lines.empty?
      init = lines.first.text_elements.inject([]) { |memo, text_chunk|
        next memo if text_chunk.text =~ ONLY_SPACES_RE
        memo << Tabula::ZoneEntity.new(*text_chunk.tlwh)
        memo
      }

      regions = lines[1..-1]
        .inject(init) do |column_regions, line|

        line_text_elements = line.text_elements.clone.select { |te| te.text !~ ONLY_SPACES_RE }

        column_regions.each do |cr|

          overlaps = line_text_elements
            .select { |te| te.text !~ ONLY_SPACES_RE && cr.horizontally_overlaps?(te) }

          overlaps.inject(cr) do |memo, te|
            cr.merge!(te)
          end

          line_text_elements = line_text_elements - overlaps
        end

        column_regions += line_text_elements.map { |te| Tabula::ZoneEntity.new(*te.tlwh) }
      end

      regions.map { |r| r.right.round(2) }.uniq
    end

    ##
    # add a TextElement to this TextChunk
    def <<(text_element)
      self.text_elements << text_element
      self.merge!(text_element)
    end

    def merge!(other)
      if other.instance_of?(TextChunk)
        if (self <=> other) < 0
          self.text_elements = self.text_elements + other.text_elements
        else
          self.text_elements = other.text_elements + self.text_elements
        end
      end
      super(other)
    end

    ##
    # split this TextChunk vertically
    # (in place, returns the remaining chunk)
    def split_vertically!(y)
      raise "Not Implemented"
    end

    def text
      self.text_elements.map(&:text).join
    end

    def inspect
      "#<TextChunk: #{self.top.round(2)},#{self.left.round(2)},#{self.bottom.round(2)},#{right.round(2)} '#{self.text}'>"
    end

    def to_h
      super.merge(:text => self.text)
    end
  end
end

File: ext/tabula/src/org/nerdpower/tabula/TextChunk.java
package org.nerdpower.tabula;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("serial")
public class TextChunk extends Rectangle {
    List<TextElement> textElements = new ArrayList<TextElement>();
    
    public TextChunk(float top, float left, float width, float height) {
        super(top, left, width, height);
    }
    
    public TextChunk(TextElement textElement) {
        super(textElement.y, textElement.x, textElement.width, textElement.height);
        this.textElements.add(textElement);
    }
    
    public TextChunk merge(TextChunk other) {
        if (this.compareTo(other) < 0) {
            this.textElements.addAll(other.textElements);
        }
        else {
            this.textElements.addAll(0, other.textElements);
        }
        super.merge(other);
        return this;

    }
    
    public void add(TextElement textElement) {
        this.textElements.add(textElement);
        this.merge(textElement);
    }

    public List<TextElement> getTextElements() {
        return textElements;
    }
    
    public static List<Line> groupByLines(List<TextChunk> textChunks) {
        float bbwidth = Rectangle.boundingBoxOf(textChunks).width;
        List<Line> lines = new ArrayList<Line>();
        
        Line l = new Line();
        l.addTextChunk(textChunks.get(0));
        textChunks.remove(0);
        lines.add(l);

        Line last = lines.get(lines.size() - 1);
        for (TextChunk te: textChunks) {
            if (last.horizontalOverlapRatio(te) < 0.1) {
                if (last.width / bbwidth > 0.9 && allSameChar(last.getTextElements())) {
                    lines.remove(lines.size() - 1);
                }
                lines.add(new Line());
                last = lines.get(lines.size() - 1);
            }
            last.addTextChunk(te);
        }
        
        if (last.width / bbwidth > 0.9 && allSameChar(last.getTextElements())) {
            lines.remove(lines.size() - 1);
        }
        
        // TODO
        // implement lines.map!(&:remove_sequential_spaces!)
        
        return lines;
    }
    
    public String toString() {
        StringBuilder sb = new StringBuilder();
        String s = super.toString();
        sb.append(s.substring(0, s.length() - 1));
        sb.append(String.format(", text=\"%s\"]", this.getText()));
        return sb.toString();
    }
    
    public String getText() {
        StringBuilder sb = new StringBuilder();
        for (TextElement te: this.textElements) {
            sb.append(te.getText());
        }
        return sb.toString();
    }
    
    public static boolean allSameChar(List<TextChunk> textChunks) {
        StringBuilder sb = new StringBuilder();
        for (TextChunk tc: textChunks) {
            sb.append(tc.getText());
        }
        String s = sb.toString();
        char c = s.charAt(0);
        for (int i = 1; i < s.length(); c = s.charAt(i), i++) {
            if (c != s.charAt(i)) {
                return false;
            }
        }
        return true;
    }


}
